### DISCUSSION TITLE
[NET 6.0] Can osx-arm64 be published under a Windows x64 machine?

### DISCUSSION THREAD

**Discussion by sn4k3** [dotnet/runtime#66574] [Q&A]
I need clarification on this title subject.

On my machine (Windows 11 x64) i publish multiple packages of my app, now with net6.0 one of them is **osx-arm64**. But users are not able to run it without call `dotnet appname.dll`. In my attempts i ask an mac M1 user to publish the app from code in his machine, and it work right away by double clicking the binary.

Compile command is as follow: `dotnet publish $project -o "publish" -c Release -r osx-arm64 -p:PublishReadyToRun=true --self-contained`

My question is as title, it's possible to use a x64 machine to make the compilation for arm64? Or that's the cause of my problem?

Issue source: https://github.com/sn4k3/UVtools/issues/431#issuecomment-1066248288

**Comment by vitek-karas** [comment]
This is probably due to signing. M1 macOS requires all executables to be signed. When the app is built on macOS the SDK will call the signing tool to "Fake sign it", so that it works at least locally. When the app is built anywhere else, the singing tool is not available (and fake signing may not work anyway).

**Reply by sn4k3** [reply]
This means i'm stuck with a mac host requirement right?

**Reply by vitek-karas** [reply]
I honestly don't know - it's more a question to macOS, if it's possible to sign an executable on a non-macOS platform.
@agocke @VSadov might have more ideas...

**Reply by sn4k3** [reply]
I found this, but not sure if other tool exists to make it possible

![msedge_1JO8Abff2L](https://user-images.githubusercontent.com/113281/158179967-b9a86046-9172-4c8c-92c2-e96c0e810748.png)

Source: https://docs.avaloniaui.net/docs/distribution-publishing/macos

**Reply by PathogenDavid** [reply]
> This is probably due to signing. M1 macOS requires all executables to be signed.

This was my initial thought too, but A) I seem to recall notarization not applying for apps launched from the terminal and B) the error dialog for failed notarization explicitly says something about the app being untrusted.

> This means i'm stuck with a mac host requirement right?

I would look into using [GitHub Actions](https://docs.github.com/en/actions) to build and deploy your app. It's free for OSS projects and they have macOS runners. (Although setting up notarization in the first place might be nightmarish without a Mac of your own to do the initial setup.)

**Reply by agocke** [reply]
You'll have to have a Mac for signing, and signing is definitely required for ARM64 (even at the terminal), and newer x64 versions soon, if not now. Apple doesn't make `codesign` available for any other platforms.

**Reply by sn4k3** [reply]
> I would look into using [GitHub Actions](https://docs.github.com/en/actions) to build and deploy your app. It's free for OSS projects and they have macOS runners. (Although setting up notarization in the first place might be nightmarish without a Mac of your own to do the initial setup.)

I think this is my best shot, i will study that, because I hate run slow and buggy macos VM on windows, it's really a pain.

> Apple doesn't make codesign available for any other platforms.

Sad 😔

**Reply by jkoritzinsky** [reply]
There are some third-party tools that emulate the behavior of `codesign` on non-Windows, but I don't know if they'll work for your scenario.

**Reply by sn4k3** [reply]
> There are some third-party tools that emulate the behavior of codesign on non-Windows, but I don't know if they'll work for your scenario.

If linux then it's ok, i already use WSL on my publish script.
Can you indicate those tools?

**Reply by jkoritzinsky** [reply]
From a quick Google search, I've found the following tools:

https://lib.rs/crates/apple-codesign

https://github.com/zhlynn/zsign

I have absolutely no idea if they work, but maybe one of them will do what you want.

**Reply by sn4k3** [reply]
Will try those and report back, thanks!

**Reply by Shmellyorc** [reply]
Whomever said you need a codesigning to run on mac is pretty much false information.

Yes you will need to chmod but it does run.

as proof. Those are Pizzamon.app bundle structure (the game has over 500+ files in it) not extracted, etc:

https://user-images.githubusercontent.com/5901214/159250086-abb2d029-892a-4735-9b17-7d8669449003.mp4


want more infomration? Look here: https://github.com/dotnet/runtime/issues/66941

**Reply by sn4k3** [reply]
@shmellyorc Looks like your app is x64? Does it work with a native arm bundle? My problem is with native arm (osx-arm64) build. x64 runs fine even without any chmod.

I see you use `p:TieredCompilation` does it make any diference on this senario?

**Reply by Shmellyorc** [reply]
> @shmellyorc Looks like your app is x64? Does it work with a native arm bundle? My problem is with native arm (osx-arm64) build. x64 runs fine even without any chmod.
> 
> I see you use `p:TieredCompilation` does it make any diference on this senario?

Reason why I use p:TieredCompilation is because of Monogame framework. It is required so it doesn't stutter. Outside of Monogame, a normal dotnet core 6 arm-x64 console app, I don't know :(.

Pretty much I built this on my PC not mac so I had to bundle it myself but my build process is:
dotnet publish -c Release -r osx-x64 /p:PublishReadyToRun=false /p:TieredCompilation=false --self-contained

**Reply by sn4k3** [reply]
I also bundled everything on PC WSL, as said, the osx-x64 works on both x64 and arm64(emulated), but osx-arm64 dont work with macOS arm64 when it come from web download.
I haven't try codesign yet because i own no apple developer account, but most likely is the solution.

My script: https://github.com/sn4k3/UVtools/blob/master/build/createRelease.sh

**Reply by agocke** [reply]
Notarization is stricter for ARM64 than for x64. Given that Apple has been escalating notarization requirements in most of its OS updates, on both x64 and ARM64, it's possible that a future update will stop your app from working without notarization on x64 as well.

**Reply by Shmellyorc** [reply]
> Notarization is stricter for ARM64 than for x64. Given that Apple has been escalating notarization requirements in most of its OS updates, on both x64 and ARM64, it's possible that a future update will stop your app from working without notarization on x64 as well.

Well.. Mac going to lose a bunch of customers than. I know Mac is trying to make themselves appear "exclusive" with the ARM64 but it'll be a cost of their doing. exclusivity == no competition == more money

 All I do is make games so I don't need to worry about signing since Steam covers the code signing. Feel sorry for the people making apps outside of steam as majority of your expenses are from Code Signing. I  know from code signing, you are paying a massive chunk of money for the app being certified and pretty much saying "its a good app". I cannot support code signing because, to me, feels like a monopoly.

Pretty much: "Get code signed or your app wont work here any more." What happened being creative and designing new cool apps to make peoples lives better? I know many developers don't finish their app because the massive amount of cash required for code signing or other certificates.

Kind of sad... but I hope you release your awesome apps.

-shmelly
